cli/src/main.rs

```diff
@@ -1,19 +1,561 @@
+        }
+        Commands::Stop { domain, rpc_port } => {
+            let _ = stop(domain, rpc_port).await;
+        }
+        Commands::Deploy {
+            domain,
+            http_port,
+            rpc_port,
+            component,
+        } => {
+            let _ = deploy(domain, http_port, rpc_port, component).await;
+        }
+    }
+
+    Ok(())
+}
+
+fn build() -> Result<(), DynError> {
+    let cargo = env::var("CARGO").unwrap_or_else(|_| "cargo".to_string());
+    let status = std::process::Command::new(cargo)
+        .current_dir(project_root())
+        .args(&["xtask", "build"])
+        .status()?;
+
+    if !status.success() {
+        Err(format!(
+            "`cargo xtask build` failed.
+
+Status code: {}",
+            status.code().expect("Build failed: no status")
+        ))?;
+    }
+
+    Ok(())
+}
+
+// We use the tonic crate to send an EchoRequest to the development_server through a gRPC address
+// The `just_started` argument is used to return ServerState::StartingUp in stead of
+// ServerState::NotStarted when "transport error" is returned by the gRPC client.
+async fn server_state(address: String, just_started: &bool) -> Result<ServerState, ServerError> {
+    let payload = "cli::server_state()".to_string();
+    match DevelopmentClient::connect(address).await {
+        Ok(mut client) => {
+            let message = EchoRequest {
+                message: payload.clone(),
+            };
+            let request = tonic::Request::new(message);
+            let response = client.echo(request).await;
+
+            match response.unwrap().into_inner() {
+                EchoReply { message } => {
+                    if message == payload.to_string() {
+                        warn!("Development server already listening");
+                        return Ok(ServerState::Started);
+                    } else {
+                        error!(
+                            "Unexpected EchoReply message from RPC server. Message: {}",
+                            message
+                        );
+                    }
+                    return Ok(ServerState::StartingUp);
+                }
+            };
+        }
+        Err(err) => match err.to_string().as_str() {
+            "transport error" => {
+                return match *just_started {
+                    true => Ok(ServerState::StartingUp),
+                    false => Ok(ServerState::NotStarted),
+                };
+            }
+            err => {
+                return Err(ServerError::ServerError { cause: err.into() });
+            }
+        },
+    };
+}
+
+async fn poll_server_state(
+    domain: &str,
+    rpc_port: &u16,
+    just_started: &bool,
+) -> Result<ServerState, PollError> {
+    let start = Instant::now();
+    let timeout = Duration::from_secs(10);
+
+    loop {
+        let rpc_addr = format!("http://{}:{}", domain, rpc_port);
+        let state = server_state(rpc_addr, just_started).await;
```

Comment:
no-op here

Just bringing this to your attention for future Tokio hacking purposes 

https://docs.rs/tokio/latest/tokio/time/fn.timeout.html

Reply:
Wondering what you mean by no-op. When I println! the value of `state` it is in fact looping till the development_state is Echo'able:

```terminal
     Running `target/debug/cli start`
[2023-09-26T13:27:46Z INFO  cli] Starting development server
[2023-09-26T13:27:46Z INFO  cli] Started development server
🪵 [main.rs:252]~ token ~ state = StartingUp
🪵 [main.rs:252]~ token ~ state = StartingUp
[2023-09-26T13:27:48Z WARN  cli] Development server already listening
🪵 [main.rs:252]~ token ~ state = Started
[2023-09-26T13:27:48Z INFO  cli] You can reach the development server on http://127.0.0.1:3001
[2023-09-26T13:27:48Z INFO  cli] handle_stderr:     Finished dev [unoptimized + debuginfo] target(s) in 0.33s
```